packages/edit-navigation/src/store/actions.js

```diff
@@ -49,171 +39,380 @@ export function setSelectedMenuId( menuId ) {
-		yield dispatch( noticesStore, 'createErrorNotice', errorMessage, {
+			: __( 'Unable to save: An error o1curred.' );
+		registry.dispatch( noticesStore ).createErrorNotice( errorMessage, {
 			type: 'snackbar',
 		} );
+	} finally {
+		registry.dispatch( coreDataStore ).__unstableReleaseStoreLock( lock );
 	}
-} );
+};
 
-function mapMenuItemsByClientId( menuItems, clientIdsByMenuId ) {
-	const result = {};
-	if ( ! menuItems || ! clientIdsByMenuId ) {
-		return result;
-	}
-	for ( const menuItem of menuItems ) {
-		const clientId = clientIdsByMenuId[ menuItem.id ];
-		if ( clientId ) {
-			result[ clientId ] = menuItem;
+/**
+ * Converts a post into a flat list of menu item entity records,
+ * representing the desired state after the save is finished.
+ *
+ * @param {Object}   post         The post.
+ * @param {Object[]} oldMenuItems The currently stored list of menu items.
+ * @return {Function} An action creator
+ */
+const getDesiredMenuItems = ( post, oldMenuItems ) => ( { dispatch } ) => {
+	const entityIdToBlockId = dispatch(
+		getEntityRecordIdToBlockIdMapping( post.id )
+	);
+
+	const blockIdToOldEntityRecord = {};
+	for ( const oldMenuItem of oldMenuItems ) {
+		const blockId = entityIdToBlockId[ oldMenuItem.id ];
+		if ( blockId ) {
+			blockIdToOldEntityRecord[ blockId ] = oldMenuItem;
 		}
 	}
-	return result;
-}
 
-function* batchSave( menuId, menuItemsByClientId, navigationBlock ) {
-	const { nonce, stylesheet } = yield apiFetch( {
-		path: '/__experimental/customizer-nonces/get-save-nonce',
-	} );
-	if ( ! nonce ) {
-		throw new Error();
+	const blocksList = blocksTreeToFlatList( post.blocks[ 0 ].innerBlocks );
+	return blocksList.map( ( { block, parentBlockId }, idx ) =>
+		blockToMenuItem(
+			block,
+			blockIdToOldEntityRecord[ block.clientId ],
+			blockIdToOldEntityRecord[ parentBlockId ]?.id,
+			idx,
+			post.meta.menuId
+		)
+	);
+};
+
+/**
+ * A selector in disguise. It returns mapping between menu item ID and it's related blocks client id.
+ *
+ * @param {number} postId The id of the stub post to get the mapping for.
+ * @return {Function} An action creator
+ */
+const getEntityRecordIdToBlockIdMapping = ( postId ) => ( { registry } ) =>
+	registry.stores[ STORE_NAME ].store.getState().mapping[ postId ] || {};
+
+/**
+ * Persists the desiredEntityRecords while preserving IDs from oldEntityRecords.
+ * The batch request contains the minimal number of requests necessary to go from
+ * desiredEntityRecords to oldEntityRecords.
+ *
+ * @param {string}   kind                 Entity kind.
+ * @param {string}   type                 Entity type.
+ * @param {Object[]} oldEntityRecords     The entity records that are currently persisted.
+ * @param {Object[]} desiredEntityRecords The entity records are to be persisted.
+ * @return {Function} An action creator
+ */
+const batchSaveChanges = (
+	kind,
+	type,
+	oldEntityRecords,
+	desiredEntityRecords
+) => async ( { dispatch, registry } ) => {
+	const changeset = dispatch(
```

Comment:
Why have this intermediary `changeset` array? Can't `batchSaveChanges` build the list of tasks given to `__experimentalBatch` itself?

Reply:
Studying it more closely, I suppose that it's to allow `batchSaveChanges` to be written in a generic way such that it could live in `@wordpress/core-data`. I get that, but... YAGNI? 🙂

In my experience it's better to introduce a new abstraction too late than too early as doing the latter can lead to codebases that have a lot of indirection.

Up to you. I'd feel better about this approach if `saveWidgetArea` used it too.

Reply:
Being generic is nice, but the real purpose is to encapsulate `getFailedChanges` logic – it needs to reason differently about saves and deletes.

In my perfect world, we wouldn't need to call `getLastEntitySaveError` or anything like that because `__experimentalBatch` would throw something actionable on failure, but that's beyond the scope of this PR. Once we get there, though, we could simplify this even more.

> Up to you. I'd feel better about this approach if saveWidgetArea used it too.

This is what I'm building up to! The truth is that we could use the same structure in any editor where you manipulate entities through blocks.

Reply:
I trust you!